[docs] Migration guide (#3896)
kmagiera/react-native-gesture-handler: packages/docs-gesture-handler/docs/guides/upgrading-to-3.mdx

---
id: upgrading-to-3
title: Upgrading to the new API introduced in Gesture Handler 3
---

## Migrating gestures

import CodeComparison from '@site/src/components/CodeComparison'; 

> The most important change brought by the Gesture Handler 3 is the new hook API. Migration is pretty straight-forward. Instead of calling builder methods, everything is passed into configuration object.

Reviewer: ```suggestion
The most important change brought by the Gesture Handler 3 is the new hook API. Migration is pretty straightforward. Instead of calling builder methods, everything is passed as a configuration object.
```

<CodeComparison
label1={"RNGH 2"}
label2={"RNGH 3"}
code1={
`const gesture = Gesture.Pan()
  .onBegin(() => {
    console.log('Pan!');
  })
  .minDistance(25);
`}
code2={
`const gesture = usePanGesture({
      onBegin: () => {
        console.log('Pan!');
      },
      minDistance: 25,
    });
`}
/>

:::danger Force Touch Gesture
[`ForceTouch`](/docs/2.x/gestures/force-touch-gesture) gesture is not available in hook API.

> :::

Reviewer: Is `ForceTouch` important enough to deserve a danger box? It focuses readers attention on it, but the feature was very niche, so almost nobody will need this info,

const pan2 = usePanGesture({
  requireToFail: pan1,
});
`}
skipFormatting1={true}
/>

Full changes are as follows:

| RNGH2                                     | RNGH3                       |
| ----------------------------------------- | --------------------------- |
| `gesture.simultaneousWithExternalGesture` | `gesture.simultaneousWith`  |
| `gesture.requireExternalGestureToFail`    | `gesture.requireToFail`     |
| `gesture.blocksExternalGesture`           | `gesture.block`             |

## Using SVG

In Gesture Handler 2 it was possible to use `GestureDetector` directly on `SVG`. In Gesture Handler 3, the correct way to interact with `SVG` is to use `InterceptingGestureDetector` and `VirtualGestureDetector`.

<CodeComparison
label1={"RNGH 2"}
label2={"RNGH 3"}
code1={
`
<GestureDetector gesture={containerTap}>
  <Svg>
    <GestureDetector gesture={circleTap}>
      <Circle />
    </GestureDetector>
  </Svg>
</GestureDetector>
`}
code2={
`
<InterceptingGestureDetector gesture={containerTap}>
  <Svg>
    <VirtualGestureDetector gesture={circleTap}>
      <Circle />
    </VirtualGestureDetector>
  </Svg>
</InterceptingGestureDetector>
`}
/>

:::danger Detectors order

> `VirtualGestureDetector` has to be a child of `InterceptingGestureDetector`.

Author: [80f4d06](https://github.com/software-mansion/react-native-gesture-handler/pull/3896/commits/80f4d06bf8d2beb2fde466e967abfdf5404d69b0)
